net/ip.c

```diff
@@ -123,29 +130,31 @@ static int m0_net_ip_inet_parse(const char *name, struct m0_net_ip_addr *addr)
 				break;
 		}
 	}
+	if (s >= ARRAY_SIZE(ip_protocol))
+		return M0_ERR(-EINVAL);
 	if (ep_name[shift] != ':')
 		return M0_ERR(-EINVAL);
 	ep_name += shift + 1;
 	at = strchr(ep_name, '@');
-	if (at == NULL) {
+	if (at == NULL)
 		return M0_ERR(-EINVAL);
-	} else {
+	else {
 		at++;
-		if (at == NULL)
+		if (at == NULL || at[0] < '0' || at[0] > '9')
 			return M0_ERR(-EINVAL);
-		memcpy(port, at, (strlen(at)+1));
+		strcpy(port, at);
 		portnum = atoi(port);
 		M0_ASSERT(portnum < M0_NET_IP_PORT_MAX);
 		addr->na_port = (uint16_t)portnum;
 	}
 
 	n = m0_net_hostname_to_ip((char *)ep_name, ip, addr);
-	if (n > 0)
-		memcpy(node, ip, n);
-	else
-		return n == 0 ? M0_ERR(-EPROTO) : M0_ERR(n);
+	if (n == 0)
+		M0_ERR(-EPROTO);
+	else if (n < 0)
+		M0_ERR(n);
 
-	inet_pton(AF_INET, node, &addr->na_n.sn[0]);
+	inet_pton(AF_INET, ip, &addr->na_n.sn[0]);
```

Comment:
Add todo here to select between IPv4 and IPv6.

Reply:
Added